src/pages/Synchronization/Programs/ProgramQueries.js

```diff
@@ -0,0 +1,100 @@
+import { useDataQuery } from '@dhis2/app-runtime'
+import { useEffect, useState } from 'react'
+import { filterListByReadAccess } from '../../../utils/utils'
+import { useWorkflowContext } from '../../../workflow-context'
+
+/**
+ * Combine multiple arrays of objects and group them by program id
+ * returns Grouped object by program id
+ */
+const groupByProgramId = (...arrays) => {
+    return arrays.flat().reduce((acc, item) => {
+        const programId =
+            typeof item?.program === 'string'
+                ? item?.program
+                : item?.program?.id
+
+        if (!programId) {
+            return acc
+        }
+
+        if (!acc[programId]) {
+            acc[programId] = []
+        }
+
+        acc[programId].push(item)
+
+        return acc
+    }, {})
+}
+
+/**
+ * Combines a list of programs with their corresponding filters
+ * returns A new array of programs including filters where applicable
+ */
+const combineProgramsWithFilters = (programsList, filtersByProgramId) => {
+    if (
+        !Array.isArray(programsList) ||
+        typeof filtersByProgramId !== 'object'
+    ) {
+        return []
+    }
+
+    return programsList.map((program) => {
+        const filters = filtersByProgramId[program.id] || []
+        return filters.length ? { ...program, filters } : { ...program }
```

Comment:
I feel like it would probably simplify things to just return `{...program, filters}`, then filters would be an empty array in case there being no matching filters, but you've also added logic checks in the consuming code, so it's not necessary.